File: data_structures/week2/make_heap/build_heap.go
package main

import "fmt"

var swaps [][]int

func shiftDown(i int, a []int) {
	min := i
	leftChild := 2*i + 1
	if leftChild < len(a) && a[leftChild] < a[min] {
		min = leftChild
	}
	rightChild := 2*i + 2
	if rightChild < len(a) && a[rightChild] < a[min] {
		min = rightChild
	}
	if min != i {
		swaps = append(swaps, []int{i, min})
		a[i], a[min] = a[min], a[i]
		shiftDown(min, a)
	}
}

func builder(a []int) {
	for i := len(a)/2 - 1; i >= 0; i-- {
		shiftDown(i, a)
	}
}

func main() {
	a := []int{5, 4, 3, 2, 1}
	builder(a)
	fmt.Println(len(swaps))
	for _, swap := range swaps {
		fmt.Println(swap[0], swap[1])
	}
}

File: data_structures/week2/make_heap/build_heap.py
# python3

class HeapBuilder:
  def __init__(self):
    self._swaps = []
    self._data = []

  def ReadData(self):
    n = int(input())
    self._data = [int(s) for s in input().split()]
    assert n == len(self._data)

  def WriteResponse(self):
    print(len(self._swaps))
    for swap in self._swaps:
      print(swap[0], swap[1])

  def ShiftDown(self, i):
    minimum = i
    a = self._data

    leftChild = 2*i + 1
    if leftChild < len(a) and a[leftChild] < a[minimum]:
      minimum = leftChild

    rightChild = 2*i + 2
    if rightChild < len(a) and a[rightChild] < a[minimum]:
      minimum = rightChild

    if minimum != i:
      self._swaps.append([i,minimum])
      a[i], a[minimum] = a[minimum], a[i]
      self.ShiftDown(minimum)

  def GenerateSwaps(self):
    i = len(self._data)//2-1
    while i>=0:
      self.ShiftDown(i)
      i=i-1

  def Solve(self):
    self.ReadData()
    self.GenerateSwaps()
    self.WriteResponse()

if __name__ == '__main__':
    heap_builder = HeapBuilder()
    heap_builder.Solve()
